### ISSUE TITLE
Next.js 14 - The server could not finish this Suspense boundary

### ISSUE THREAD

**Issue by Thanaen** [PostHog/posthog-js#872]
Hello,
I followed [the documentation](https://posthog.com/docs/libraries/next-js) to the letter to set up Posthog with Next.js, and in production I get this error:
```
Uncaught Error: Minified React error #419; visit https://reactjs.org/docs/error-decoder.html?invariant=419 for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
```

According to the React.js documentation, it goes like this:
`The server could not finish this Suspense boundary, likely due to an error during server rendering. Switched to client rendering.`

It looks like Next.js doesn't like me wrapping the "PostHogPageView" component in <Suspense>. 😅

I'm using Next.js 14.0.1, and I don't get the error when I'm in development mode.

**Comment by Arsenalist** [comment]
The basic example provided for NextJS does not work unless you remove the suspense boundary. Easily reproducible - just follow the directions in the docs.

**Comment by Arsenalist** [comment]
@Thanaen did you ever figure this out?

**Comment by Thanaen** [comment]
Unfortunately no, there has been no news on the open issue at Next.js (https://github.com/vercel/next.js/issues/51581)

**Comment by Arsenalist** [comment]
Thanks, for now I just capture the events automatically and got rid of the manual capture. Good enough.

**Comment by asadeghi** [comment]
I'm seeing the same error with Next 13.5.6 as well as 14.0.4.

**Comment by maybebansky** [comment]
@Arsenalist If I remove the suspense boundary I don't get this error on page-load and events seem to be being captured OK. Any idea what you're losing though? I assume it wouldn't be in the official docs for no reason.

**Comment by Thanaen** [comment]
> @Arsenalist If I remove the suspense boundary I don't get this error on page-load and events seem to be being captured OK. Any idea what you're losing though? I assume it wouldn't be in the official docs for no reason.

It seems to me that calling a component that uses the "usePathname" hook without enclosing it in a Suspense tag causes the parent to opt for client rendering.

**Comment by asadeghi** [comment]
The other workaround @maybebansky , @Thanaen , is to remove all the manual page view capturing in `PostHogPageview` altogether, and instead set `capture_pageview` param in `posthog.init` to true to opt for automatic pageview capture. PostHogPageview calls `posthog.capture` directly, but you can alternatively just allow Posthog to do it automatically with this flag.

**Comment by maybebansky** [comment]
> The other workaround @maybebansky , @Thanaen , is to remove all the manual page view capturing in `PostHogPageview` altogether, and instead set `capture_pageview` param in `posthog.init` to true to opt for automatic pageview capture. PostHogPageview calls `posthog.capture` directly, but you can alternatively just allow Posthog to do it automatically with this flag.

This does seem to work. Are you aware of any downside? 

In my dashboard I can see the page url and search param is being logged correctly, so im not sure why the official docs would suggest using the extra code for the PostHogPageview component?

**Comment by maybebansky** [comment]
> The other workaround @maybebansky , @Thanaen , is to remove all the manual page view capturing in `PostHogPageview` altogether, and instead set `capture_pageview` param in `posthog.init` to true to opt for automatic pageview capture. PostHogPageview calls `posthog.capture` directly, but you can alternatively just allow Posthog to do it automatically with this flag.

I think this only captures the first page load, but not subsequent navigations. This is covered in the docs: https://posthog.com/tutorials/nextjs-analytics#capturing-pageviews-in-nextjs

Ive also found it be true in my set up.

**Comment by maybebansky** [comment]
Not sure if its a good solution but only loading the PostHogPageview component after a useEffect  has run also seems to work 

```
export function PostHogPageviewLoader() {
  const [loaded, setLoaded] = useState(false);
  useEffect(() => {
    setLoaded(true);
  }, []);
  if (!loaded) return null;
  return <PostHogPageview />;
}
```

**Comment by gruckion** [comment]
I am getting this error.

```
Error:   x `ssr: false` is not allowed with `next/dynamic` in Server Components. Please move it into a client component.
```

Now they don't let you use ssr false with dynamic in the server component. So unless we change the layout into a client component then we can't use dynamic?

**Comment by ivanagas** [comment]
@gruckion with Next.js 15, you need to use a `Suspense` again, we updated our instructions here: https://posthog.com/docs/libraries/next-js#router-specific-instructions

**Comment by zachkirsch** [comment]
In Next.js 15, I am getting this error unles I remove the suspense:
```
Uncaught Error: Minified React error #419; visit https://reactjs.org/docs/error-decoder.html?invariant=419 for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
```

**Comment by LTe** [comment]
I have the same problem on `Next.js v15.1.6` in production mode. There is no error in development.

**Comment by ivanagas** [comment]
@LTe can you share a reproduction? 

- App or pages router?
- What does your `providers` file look like?
- What does your `layout` file look like? 
- What does your `pageview` file look like?

**Comment by LTe** [comment]
@ivanagas yes — https://github.com/LTe/posthog-suspense-error-on-production

**Steps to reproduce**

1. Install dependencies `yarn install`
2. Build next application `yarn build`
3. Start production next application `yarn start`
4. Visit `localhost:3000`

**Expected Behavior**

There is no errors in console

**Actual Behavior**

```
Uncaught Error: Minified React error #419; visit https://react.dev/errors/419 for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
```

**Comment by ivanagas** [comment]
In `layout.tsx`, only making the `<PostHogProvider>` wrap the `{children}`, not the `<body>` seems to fix this. This is already what we have in the [docs](https://posthog.com/docs/libraries/next-js#router-specific-instructions).

```
// app/layout.tsx
import "./globals.css";
import { PostHogProvider } from "./providers";

export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <html lang="en">
        <body>
          <PostHogProvider>
            {children}
          </PostHogProvider>
      </body>
    </html>
  );
}
```

**Comment by LTe** [comment]
@ivanagas you are right, when `PostHogProvider` is inside body there is no problem. I guess this is more NextJS issue. Thanks for looking into this! <3

**Comment by Danielvandervelden** [comment]
> Not sure if its a good solution but only loading the PostHogPageview component after a useEffect has run also seems to work
> 
> ```
> export function PostHogPageviewLoader() {
>   const [loaded, setLoaded] = useState(false);
>   useEffect(() => {
>     setLoaded(true);
>   }, []);
>   if (!loaded) return null;
>   return <PostHogPageview />;
> }
> ```

Thanks for this. I've been pulling my hair out. It's stupid that we need such a solution, but it's a solution none-the-less.
